cpp/src/arrow/array/array_nested.cc

```diff
@@ -836,24 +845,31 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
 
   using OffsetArrayType = typename TypeTraits<OffsetArrowType>::ArrayType;
   const auto& typed_offsets = checked_cast<const OffsetArrayType&>(*offsets);
-  auto buffers = BufferVector({nullptr, typed_offsets.values()});
+
+  if (null_bitmap != nullptr) {
+    auto buffers = BufferVector({std::move(null_bitmap), typed_offsets.values()});
+    return std::make_shared<MapArray>(type, offsets->length() - 1, std::move(buffers),
+                                      keys, items, /*null_count=*/null_bitmap->size(),
+                                      offsets->offset());
+  }
+
+  auto buffers = BufferVector({null_bitmap, typed_offsets.values()});
   return std::make_shared<MapArray>(type, offsets->length() - 1, std::move(buffers), keys,
                                     items, /*null_count=*/0, offsets->offset());
```

Comment:
Small nit: I think it will be easier to read the code later if you only define the buffers (and null_count) in an if / else, and the reuse the same `return std::make_shared<MapArray>(..)` (then you don't need to wonder what is exactly the difference between both)